src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -101,6 +101,14 @@ def make_install_req_from_dist(dist, parent):
     return ireq
 
 
+def is_already_installed(cand):
+    # type: (Candidate) -> bool
+    # For an ExtrasCandidate, we check the base
+    if isinstance(cand, ExtrasCandidate):
+        cand = cand.base
+    return isinstance(cand, AlreadyInstalledCandidate)
```

Comment:
I wonder if it’s better to add an `is_installed()` on candidates.

Reply:
We can't abstract out `keep_installed()` out since it depends on the `upgrade_strategy`.

+1 on adding `is_installed()` -- although, as a read-only property feels a lot cleaner.

Reply:
I considered that, but wasn't sure if it was too intrusive (it felt a bit specialised). But as both of you seem OK with the idea, I'll do this.

Reply:
OK. I made `is_installed` an attribute on the class, and added an equivalent property to the `Candidate` class. That seems to work, although mypy failed to catch that I forgot to add the attribute to `RequiresPythonCandidate`. I understand why (there's a default implementation in `Candidate`) but is there really no way to say in mypy "every subclass of this class needs to have an attribute `is_installed`"? Or to say "every object of type `Candidate` must have an `is_instance` attribute" and have class attributes count?

Reply:
I think that’s what Protocol is for, but it’s a pain to use Protocol on comment-style annotations. Python 2 😞